### ISSUE TITLE
Padding problem with short utterances with Wav2Vec2 

### ISSUE THREAD

**Issue by RobinAlgayres** [pytorch/fairseq#3118]
#### What is your question?
Let's consider two types of inputs:
1) 100ms of speech at 16kHz: [t0,...,tn]
2) this very same speech file but with zero padding at the end: [t1,....,tn,0,0,...,0,0]
 
I give both sequences to Wav2vec2 (in eval mode) and I get completely different kind of features. 
My question is: How to pad short utterances so that the zero padding does not impact the wav2vec2 features ? 
(I need this padding to finetune wav2vec2 on a specific dataset of variable length utterances)

The code below shows that the sum of the features extracted from 1) are differents than for 2)
(I applied a masking in the second case).
Note that using  CPC, I get what I want: both extracted features are identical, meaning that this problem is specific to Wav2vec2.
 

#### Code

import torch
import fairseq
from torch import nn

model, cfg, task = fairseq.checkpoint_utils.load_model_ensemble_and_task(['wav2vec_small.pt'])
model = model[0]
model.eval()

winput = torch.randn(1,3000) # no padding mode
c = model(winput,mask=False,features_only=True)['x'].squeeze()
s=torch.sum(torch.abs(c),dim=1)
print(torch.sum(s)) # here I sum the embedding value and get 1654.80

z = torch.zeros(1,3000) # padding mode
winput=torch.cat((winput,z),dim=1)
c = model(winput,mask=False,features_only=True)['x'].squeeze()
mask=torch.ones(c.size(0))
mask[9:]=0
mask=mask.reshape(-1,1)
c=c*mask
s=torch.sum(torch.abs(c),dim=1)
print(torch.sum(s))    # here I sum the embedding value and get 1577.70

#### What's your environment?

 - fairseq Version 1.0
 - PyTorch Version 1.7
 - Linux
 - How you installed fairseq : source
 - Python version: 3.7
 - CUDA/cuDNN version: 10.0

**Comment by anogkongda** [comment]
a suggestion is to
add padding mask, a sample code is in
fairseq/data/audio/raw_audio_dataset.py: line 81 - 91

**Comment by RobinAlgayres** [comment]
Hi,
Well it does not really solve the problem. I tried giving a padding_mask to the model and the features embeddings are still quite different. Here is the code using a padding mask.

Basically I create a wav segment, I pad it with zeros, it goes through the convolutional encoder, it gets multiplied by the padding mask, go through the context network, and then I re-multiply it by the mask to show that the output is different than my previous experiment with no padding. 

winput = torch.randn(1,3000)
z = torch.zeros(1,3000)                                                                                                                winput=torch.cat((winput,z),dim=1)                                                                                                     mask=torch.ones((1,18)) # 18 is the expected size
mask[:,9:]=0 # half of the input is actual speech
pad = mask < 1 # padding mask must be boolean
c = model(winput,padding_mask=pad,mask=False,features_only=True)['x'].squeeze()
c=c*(mask.reshape(-1,1)) 
s=torch.sum(torch.abs(c),dim=1)
print(torch.sum(s))

The thing is in Wav2Vec2 there is no difference between the context network and the predictor (as it is the case in CPC), which means that I do not think it is possible to solve the issue of short utterance embedding... 
Any ideas?

**Comment by alexeib** [comment]
the way you are constructing padding mask seems to set the padded timesteps to False and non-padded to True, but in fairseq these masks work the other way (True for padded timesteps, False otherwise)

**Comment by RobinAlgayres** [comment]
I actually set to True the padded timesteps.
Here is a print of my padded mask for the tensor 'winput':

tensor([[False, False, False, False, False, False, False, False, False,  True,
          True,  True,  True,  True,  True,  True,  True,  True]])

**Comment by abarcovschi** [comment]
Hello @RobinAlgayres are there any updates on this issue? Did you get it working?
